### PR TITLE
8298187: (fs) BsdFileAttributeViews::setTimes does not support lastAccessTime on HFS+

### PR DESCRIPTION
**Pull request by bplb**
If the path is on an HFS store, use the generic Unix implementation of `java.nio.file.attribute.BasicFileAttributeView::setTimes`to set the last access time.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298187](https://bugs.openjdk.org/browse/JDK-8298187): (fs) BsdFileAttributeViews::setTimes does not support lastAccessTime on HFS+

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11600/head:pull/11600` \
`$ git checkout pull/11600`

Update a local copy of the PR: \
`$ git checkout pull/11600` \
`$ git pull https://git.openjdk.org/jdk pull/11600/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11600`

View PR using the GUI difftool: \
`$ git pr show -t 11600`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11600.diff">https://git.openjdk.org/jdk/pull/11600.diff</a>

</details>

### CODE REVIEW

**Comment by bplb** [comment]
The macOS system call `setattrlist(2)` fails to set the last access time supplied as a parameter when the file in question is on an HFS store. It appears to override the parameter value with the current system time at the moment the system call is invoked.

The portion of the proposed change starting at line 42 handles the common case where the creation time is _not_ being set. The other more substantial changes further down handle the likely uncommon case where the creation time _is_ being set.

**Comment by AlanBateman** [comment]
I don't think we should put in this workaround as it requires getting the FileStore each time that the last access time is set. This seems like a macOS bug that should be reported to Apple first.

**Comment by bplb** [comment]
> I don't think we should put in this workaround as it requires getting the FileStore each time that the last access time is set. This seems like a macOS bug that should be reported to Apple first.

I agree about reporting it to Apple, but I commented above because the changes starting at line 42 probably would work around most cases. The second group of changes could be removed.

**Comment by AlanBateman** [comment]
> I agree about reporting it to Apple, but I commented above because the changes starting at line 42 probably would work around most cases and do not need the `FileStore`. The second group of changes could be removed.

APFS became the default file system in macOS 10.13 so I don't know how common it is to see HFS+ these days. Although we expect it should be rare to be changing the last access time, my concern is that finding the mount point to get the file system type can be problematic in some configurations. It would be good if we could avoid the workaround so I'm interested to see if Apple will fix the bug or not.

**Comment by AlanBateman** [comment]
Removing the search of the mount points is good but we still have the issue that calling setTimes with a file creation time + another will be two syscalls. The spec for setTimes sets the expectation that the method is not atomic when setting some but not all timestamps, this is because it has to allow for implementation that needs to read the existing timestamps in order to update. The implementation change proposed here may require relaxing the spec Could you look into using file.openForAttributeAccess(followLinks) and futimes + fsetattrlist so the file is opened only once?

**Comment by bplb** [comment]
> Could you look into using file.openForAttributeAccess(followLinks) and futimes + fsetattrlist so the file is opened only once?

So changed in ad14b579cb865ee58e43c6fb4a35a779bfe4a578.

File: src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java [hunk 1]
```diff
@@ -82,26 +82,24 @@ private static native int clonefile0(long srcAddress, long dstAddress,
...
-                             createTime, options);
-            } finally {
-                Blocker.end(comp);
+        long comp = Blocker.begin();
+        try {
+            fsetattrlist0(fd, commonattr, modTime, accTime,
+                          createTime, options);
+        } finally {
+            Blocker.end(comp);
             }
```
Comment (AlanBateman) [inline src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java:98]: Can you check the formatting here, the end brace seems to be in pushed out.

File: src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java [hunk 2]
```diff
@@ -49,28 +49,71 @@ private static void setTimes(UnixPath path, FileTime lastModifiedTime,
...
+        int fd = -1;
         try {
-            setattrlist(path, commonattr, modValue, accValue, createValue,
-                        followLinks ?  0 : UnixConstants.FSOPT_NOFOLLOW);
-        } catch (UnixException x) {
-            x.rethrowAsIOException(path);
+            try {
+                fd = path.openForAttributeAccess(followLinks);
+            } catch (UnixException x) {
+                x.rethrowAsIOException(path);
```
Comment (AlanBateman) [inline src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java:57]: The try-open-catch shouldn't be nested in the other try-catch. If the open fails then the fd isn't closed and there is no suppressed exception.

**Review by AlanBateman — COMMENTED** [review]
Updated version looks quite good, just a few minor nits.

File: src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java (hunk 1 above)
Comment (bplb) [reply src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java]: Corrected in BsdFileAttributeViews-setTimes-8298187.

File: src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java (hunk 2 above)
Comment (bplb) [reply src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java]: Corrected in BsdFileAttributeViews-setTimes-8298187.

**Review by AlanBateman — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java [hunk 3]
```diff
@@ -49,28 +50,94 @@ private static void setTimes(UnixPath path, FileTime lastModifiedTime,
...
+                long accessValue= lastAccessTime.to(timeUnit);
+
+                boolean retry = false;
+                try {
+                    if (useLutimes)
+                        lutimes(path, accessValue, modValue);
+                    else
+                        futimens(fd, accessValue, modValue);
+                } catch (UnixException x) {
+                    // if futimens fails with EINVAL and one/both of the times is
```
Comment (AlanBateman) [inline src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java:100]: The comment should probably say "lutimes/futimens".

Comment (bplb) [reply src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java:100]: Comment modified in fc589d209bd60a63eb7935ced2be776e12ca223e.

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**